README.md
# Onroerend Erfgoed Abacus

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.6533854.svg)](https://doi.org/10.5281/zenodo.6533854)

Scripts om erfgoed uit de [Inventaris Onroerend
Erfgoed](https://inventaris.onroerenderfgodd.be) correct te helpen tellen.

## Type tellen

Het script *type_tellen* kan gebruikt worden om het aantal erfgoedobjecten van 
een bepaalde erfgoedtype te tellen. Kan gecombineerd worden met types die niet 
mogen meetellen voor dingen zoals "kerken, maar geen abdijkerken" of met zaken 
zoals de rechtsgevolgen en de erfgoedwaarde. Om het script te gebruiken, stel
je een aantal parameters in:

**DISCIPLINE = None**

Ofwel *None* om aan te geven dat alle disciplines gevraagd worden, ofwel een
list van disciplines (1, 2, 3, 4). Meestal is discipline 4 (varend erfgoed) niet
relevant. Voor het tellen van zaken zoals kerken of fonteinen is het vaak
nuttig om enkel de discipline *bouwkundig erfgoed (2)* te gebruiken.

**ERFGOEDWAARDE = None**

Ofwel *None* om aan te geven dat de huidige erfgoedwaarde geen rol speelt, of
een boolean *True* of *False* om aan te geven dat we enkel erfgoedobjecten
zoeken waarvan de erfgoedwaarde nog aanwezig is of erfgoedobjecten waarvan de
erfgoedwaarde niet meer aanwezig is (weinig relevat).

**RECHTSGEVOLGEN = None**

Ofwel *None* om aan te geven dat we alle erfgoedobjecten tellen, ongeacht de
gekoppelde beschermingen, ofwel een string zoals `zonder rechtsgevolgen`, `met
rechtsgevolgen`, `beschermd`, `vastgesteld`.

**CONCEPT = None**

Ofwel *None* om aan te geven dat alle erfgoedobjecten van alle types tellen,
ofwel een id van een erfgoedtype uit de thesaurus
[ERFGOEDTYPES](https://thesaurus.onroerenderfgoed.be/conceptschemes/ERFGOEDTYPES).
Indien dit op *None* blijft staan, kan dit ongeveer de volledige databank aan
gegevens opleveren. Dit is zelden de bedoeling en zou dan moeten gecombineerd
worden met een parameter *RECHTSGEVOLGEN = `beschermd`* en *DISCIPLINE = [4]*
zodat het zoekresultaat niet te groot wordt.

**NOT_CONCEPT = []**

Ofwel een lege list om aan te geven dat er geen erfgoedtypes moeten weggelaten
worden, ofwel een list met id's van erfgoedtypes. Deze parameter werkt enkel
indien er ook een *CONCEPT* ingevuld is en kan gebruikt wordt om bv. te zoeken
naar *katholieke kerken* maar geen *abdijkerken*

### Voorbeelden

* *DISCIPLINE = [4]* en *RECHTSGEVOLGEN = `beschermd`*: al het beschermd varend
    erfgoed
* *DISCIPLINE = [2, 3]* en *RECHTSGEVOLGEN = `beschermd`* en *CONCEPT = 230* en
    *NOT_CONCEPT = [6, 52, 242]*: alle beschermde katholieke kerken die geen abdijkerk,
    begijnhofkerk of kloosterkerk zijn.
* *DISCIPLINE = [2, 3]* en *RECHTSGEVOLGEN = `beschermd`* en *CONCEPT = 514*:
    alle beschermde windmolens


HISTORY.md
# 1.0.3 (2023-05-25)

Fixed some bugs and updated to the latest version of the inventaris service.

# 1.0.2 (2022-05-10)

Some cleaning and extra documentation.

# 1.0.1 (2022-05-09)

Mucking about with Zenodo releases. Functionally identical to the previous version.

# 1.0.0 (2022-05-09)

First version, based on older code.


utils.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
'''
Nuttige functie die helpen erfgoed in de inventaris tellen.
'''

from datetime import datetime

from skosprovider_atramhasis.providers import AtramhasisProvider
from skosprovider.providers import DictionaryProvider
from skosprovider.utils import dict_dumper

import requests

from requests import RequestException

import logging

from dogpile.cache import make_region

def fetch_query(url, parameters, SSO = None, session=None):
    '''
    Fetch all data from a url until there are no more `next` urls in the Link
    header.

    :param str url: The url to fetch from
    :param dict parameters: A dict of query string parameters
    :param str SSO: OpenAM SSO token
    :rtype: dict
    '''

    if not session:
        session = requests.Session()

    data = []

    headers = {'Accept': 'application/json'}
    if SSO:
        headers['OpenAmSSOID'] = SSO

    logging.info('Opzoeken %s met %s', url, parameters)

    res = session.get(url, params=parameters, headers=headers)
    res.raise_for_status()

    aantal = None
    if res.status_code == requests.codes.ok:
        if 'Content-Range' in res.headers:
            aantal = res.headers['Content-Range'].split('/')[-1]
        else:
            aantal = 0
    logging.info('Aantal resultaten: %s', aantal)

    if not aantal:
        return []
    else:
        data.extend(res.json())

    while 'next' in res.links:
        url = res.links['next']['url']
        logging.info('Opzoeken %s', url)
        res = session.get(url, headers=headers)

        data.extend(res.json())

    for d in data:
        # Fix for a weird bug that occasionally happened
        d['self'] = d['uri'].replace('id.erfgoed.net', 'inventaris.onroerenderfgoed.be')

    return data

def _get_url(url, cache, SSO=None, session=None):

    logging.info('Opzoeken %s', url)

    if not session:
        session = requests.Session()

    def creator():
        headers = {'Accept': 'application/json'}
        if SSO:
            headers['OpenAmSSOID'] = SSO

        for i in range(5):  # Try up to 5 times to get the URI
            # Retry scenarios are: connection errors, timeouts, 5xx status codes
            try:
                response = session.get(url, headers=headers, timeout=10)
                if response.ok:
                    return response.json()
                if response.status_code == 404:
                #if 400 <= response.status_code < 500:
                    logging.warning(
                        f"{url} werd niet gevonden: "
                        f"{response.status_code} - {response.text}"
                    )
                    return False
                else:
                    logging.error(
                        f"Probleem ophalen {url}: "
                        f"{response.status_code} - {response.text}"
                    )
            except RequestException as e:
                logging.error(
                    f"Probleem ophalen {url}: {e}"
                )
        else:  # When the loop exited without a break, all retries used.
            raise Exception(f'Ik heb 5 keer geprobeerd {url} te bereiken, maar dit is niet gelukt')

    return cache.get_or_create(url, creator)

erfgoedobjecten_region = make_region().configure(
    'dogpile.cache.dbm',
    expiration_time = 60 * 60 * 24,
    arguments = {
        'filename': './erfgoedobjecten.dbm'
    }
)

def get_erfgoedobject(url, SSO = None, session = None):

    return _get_url(url, erfgoedobjecten_region, SSO, session)

erfgoedtypes = None

def get_erfgoedtypes():
    global erfgoedtypes
    if erfgoedtypes is not None:
        return erfgoedtypes
    # Keep cache in between runs of the script
    # Value is considered valid for 1 week
    erfgoedtypes = AtramhasisProvider(
        {'id': 'vioe-erfgoedtypes)'},
        base_url='https://thesaurus.onroerenderfgoed.be',
        scheme_id='ERFGOEDTYPES',
        cache_config={
            'cache.backend': 'dogpile.cache.dbm',
            'cache.expiration_time': 60 * 60 * 24 * 7,
            'cache.arguments.filename': 'erfgoedtypes.dbm'
        }
    )

    return erfgoedtypes

def is_geldig(aanduiding):
    '''
    Is een aanduidingsobject geldig of niet?
    '''
    nu = datetime.now()
    return datetime.strptime(aanduiding['geldigheid_start'], '%d-%m-%Y') <= nu \
        and (not 'geldigheid_einde' in aanduiding \
        or datetime.strptime(aanduiding['geldigheid_einde'], '%d-%m-%Y') > nu ) \

def analyseer_aanduidingen(erfgoedobject):
    '''
    Analyseer de relaties met aanduidingsobjecten om de beschermingstoestand
    van een erfgoedobject te bepalen.
    '''
    aanduidingen = [a for a in erfgoedobject['relaties'] if a['verwant']['id'] == 5 and is_geldig(a)]

    ret = {
        'aangeduid': len([a for a in aanduidingen if is_geldig(a)]),
        'beschermd': len([a for a in aanduidingen if a['bescherming']]),
        'vastgesteld': len([a for a in aanduidingen if a['vaststelling']]),
        'erfgoedlandschap': len([a for a in aanduidingen if a['aanduidingsobjecttype'] in
            ['Erfgoedlandschap']]),
        'unesco': len([a for a in aanduidingen if a['aanduidingsobjecttype'] in
            ['Unesco werelderfgoed kernzone', 'Unesco werelderfgoed bufferzone']]),
        'monument': len([a for a in aanduidingen if a['aanduidingsobjecttype'] in
            ['Beschermd monument']]),
        'sdgezicht': len([a for a in aanduidingen if a['aanduidingsobjecttype'] in
            ['Beschermd stads- of dorpsgezicht', 'Beschermd stads- of dorpsgezicht, intrinsiek', 'Beschermd stads- of dorpsgezicht, ondersteunend']]),
        'landschap': len([a for a in aanduidingen if a['aanduidingsobjecttype'] in
            ['Beschermd cultuurhistorisch landschap']]),
        'site': len([a for a in aanduidingen if a['aanduidingsobjecttype'] in
            ['Beschermde archeologische site']]),
        'overgangszone': len([a for a in aanduidingen if a['aanduidingsobjecttype'] in
            ['Overgangszone']]),
        'ongeldige beschermingen': len([a for a in erfgoedobject['relaties'] if
            a['verwant']['id'] == 5 and a['bescherming'] and not is_geldig(a)])
    }
    ret['enkel monument'] = 'ja' if \
        ret['monument'] > 0 and \
        ret ['sdgezicht'] == 0 and \
        ret['landschap'] == 0 else 'nee'

    ret['enkel sdgezicht'] = 'ja' if \
        ret['sdgezicht'] > 0 and \
        ret ['monument'] == 0 and \
        ret['landschap'] == 0 else 'nee'

    ret['enkel landschap'] = 'ja' if \
        ret['landschap'] > 0 and \
        ret ['sdgezicht'] == 0 and \
        ret['monument'] == 0 else 'nee'

    return ret

def analyseer_relaties(erfgoedobject):
    '''
    Zoek de relaties van het type omvat en is deel van van een erfgoedobject
    op.
    '''

    erfgoedobjecten = [e for e in erfgoedobject['relaties'] if e['verwant']['id'] == 4]

    ret = {
        'omvat': [e['uri'] for e in erfgoedobjecten if e['relatietype']['id'] == 7],
        'deel van': [e['uri'] for e in erfgoedobjecten if e['relatietype']['id'] == 6]
    }

    return ret

def get_id(uri):
    return uri.split('/')[-1]

def clean_relaties(erfgoedobjecten):
    '''
    Kuis de 'deel van' en 'omvat' relaties op zodat enkel de relaties
    overblijven met andere erfgoedobjecten in dezelfde lijst.
    '''
    search = [e['uri'] for e in erfgoedobjecten]
    for e in erfgoedobjecten:
        e['deel van'] = ", ".join([get_id(dv) for dv in e['deel van'] if dv in search])
        e['omvat'] = ", ".join([get_id(omv) for omv in e['omvat'] if omv in search])
    return erfgoedobjecten

def analyseer_kenmerkgroepen(erfgoedobject, concept):
    '''
    Zoek de typologie, datering en stijl op van een erfgoedobject en de
    primaire kenmerkgroep(en)
    '''
    def get_termen(kenmerk_id):
        termen = []
        for k in erfgoedobject['kenmerkgroepen']:
            for t in k['thesaurus']:
                if t['kenmerk']['id'] == kenmerk_id:
                    termen.append(t['label'])
        return list(set(termen))

    prim_kmgr = []
    if concept:
        prim_kmgr = zoek_primaire_kenmerkgroepen(erfgoedobject, concept)

    return {
        'typologie': ", ".join(get_termen(3)),
        'datering': ", ".join(get_termen(8)),
        'stijl': ", ".join(get_termen(5)),
        'primaire kenmerkgroepen': ", ".join(
            [" | ".join(t['label'] for t in k['thesaurus']) for k in prim_kmgr]
        )
    }

def zoek_primaire_kenmerkgroepen(erfgoedobject, concept):
    '''
    Zoek de kenmerkgroepen waaraan deze typologie of één van zijn kinderen is toegekend
    '''
    exp = get_erfgoedtypes().expand(concept.id)
    expanded_concepts = []
    for c in exp:
        expanded_concepts.append(get_erfgoedtypes().get_by_id(c).uri)

    allek = None
    kmgr = erfgoedobject['kenmerkgroepen']

    # Nagaan of er een "alle kenmerkgroepen" is en de typologie uit die groep
    # toekennen aan alle kenmerkgroepen
    allek = [k for k in kmgr if k['alle_kenmerkgroepen']]
    if len(allek):
        for k in kmgr:
            if not k['id'] == allek[0]['id']:
                k['thesaurus'].extend(allek[0]['thesaurus'])
            else:
                del k

    prim_kmgr = []
    for k in kmgr:
        for t in k['thesaurus']:
            if t['kenmerk']['id'] == 3 and t['uri'] in expanded_concepts:
                prim_kmgr.append(k)
                continue

    return prim_kmgr


type_tellen.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
'''
The script allows creating data files based on the Flanders inventory of 
immovable heritage (https://inventaris.onroerenderfgoed.be) to count 
heritage and how protected it is.
'''

from skosprovider_atramhasis.providers import AtramhasisProvider
from skosprovider.providers import DictionaryProvider
from skosprovider.utils import dict_dumper

import requests

import csv
import os

from datetime import datetime

from utils import fetch_query
from utils import get_erfgoedtypes
from utils import get_erfgoedobject
from utils import analyseer_aanduidingen
from utils import analyseer_kenmerkgroepen
from utils import analyseer_relaties
from utils import clean_relaties

# Alle disciplines
DISCIPLINE = None
# Alles behalve varend
#DISCIPLINE = [1,2,3]
# Enkel bouwkundig
#DISCIPLINE = [2]

# Alles ongeacht erfgoedwaarde
ERFGOEDWAARDE = None
#Alles met erfgoedwaarde
#ERFGOEDWAARDE = True
#Alles zonder erfgoedwaarde
#ERFGOEDWAARDE = False

# Alles ongeacht de rechtsgevolgen
# RECHTSGEVOLGEN = None
# Alles wat beschermd is
RECHTSGEVOLGEN = 'beschermd'

# Alles ongeacht type
# Opgelet, niet gebruiken zonder combinatie met andere parameters.
# Dit is een zeer zware vraag die uren werk kan vragen
#CONCEPT = None
# kerken
#CONCEPT = 1005
# katholieke kerken
#CONCEPT = 230
# begijnhoven
CONCEPT = 53
# eet- en drinkgelegenheden
#CONCEPT = 881
# windmolens
#CONCEPT = 514
# hondenhokken
#CONCEPT = 192
# lagere scholen
#CONCEPT = 259
# hoeven
#CONCEPT = 190
# orgels
#CONCEPT = 1780

# NOT niets, alles dus
NOT_CONCEPT = []
# NOT abdijkerken, begijnhofkerken, kloosterkerken
#NOT_CONCEPT = [6, 52, 242]

# URLs
INVENTARIS_HOST = 'https://inventaris.onroerenderfgoed.be/'
URL_ERFGOEDOBJECTEN = INVENTARIS_HOST + 'erfgoedobjecten'

os.makedirs(f'./output', exist_ok=True)

FILENAME_LOG = './output/query_%s.log' % datetime.now()

import logging
logging.basicConfig(
    format='%(asctime)s %(levelname)-8s %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S',
    filename=FILENAME_LOG,
    filemode='w',
    level=logging.INFO
)

def generate_csv(
        discipline = None, erfgoedwaarde = None,
        rechtsgevolgen = None,
        concept = None, not_concept = []
    ):

    session = requests.Session()

    query = {}

    if discipline:
        query['discipline'] = '[%s]' % (','.join([str(d) for d in discipline]))
    if erfgoedwaarde:
        query['erfgoedwaarde'] = erfgoedwaarde
    if concept:
        query['typologie'] = [concept.id]
        not_concept = [f'-{nc.id}' for nc in not_concept]
        query['typologie'].extend(not_concept)
    if rechtsgevolgen:
        query['rechtsgevolgen'] = rechtsgevolgen

    logging.info('Uit te voeren query: %s', query)

    erfgoedobjecten = fetch_query(
        URL_ERFGOEDOBJECTEN,
        query,
        session=session
    )

    export = []
    for e in erfgoedobjecten:
        logging.info('Opvragen %s met self %s', e['uri'], e['self'])
        exp = {
            'id': e['id'],
            'uri': e['uri'],
            'self': e['self'],
            'naam': e['naam'],
            'omvang': e['omvang']['naam'],
            'disciplines': "".join([d['naam'][0] for d in e['disciplines']]),
            'locatie_samenvatting': e['locatie_samenvatting']
        }
        extra_e = get_erfgoedobject(e['self'], session=session)
        exp.update({
            'erfgoedwaarde': 'ja' if extra_e['erfgoedwaarde'] else 'nee',
            'fysieke staat': extra_e['fysieke_status']['naam']
        })
        exp.update({
            'provincie': extra_e['locatie']['provincie'],
            'gemeente': extra_e['locatie']['gemeente'],
            'deelgemeente': extra_e['locatie']['deelgemeente'],
            'straat': extra_e['locatie']['straat']
        })
        exp.update(analyseer_aanduidingen(extra_e))
        exp.update(analyseer_kenmerkgroepen(extra_e, concept))
        exp.update(analyseer_relaties(extra_e))
        export.append(exp)

    export = clean_relaties(export)
    if concept:
        filename = './output/query_%s_%s_%s.csv' % (concept.id, concept.label('nl-BE').label, datetime.now())
    else:
        filename = './output/query_%s.csv' % datetime.now()
    with open(filename, 'w', newline='') as csvfile:
        fieldnames = [
            'id', 'uri', 'self', 'naam',
            'omvang', 'disciplines',
            'erfgoedwaarde', 'fysieke staat',
            'locatie_samenvatting', 'provincie', 'gemeente',
            'deelgemeente', 'straat',
            'omvat', 'deel van',
            'aangeduid', 'vastgesteld', 'beschermd', 'erfgoedlandschap', 'unesco',
            'monument', 'enkel monument',
            'sdgezicht', 'enkel sdgezicht',
            'landschap', 'enkel landschap',
            'site', 'overgangszone',
            'ongeldige beschermingen',
            'typologie', 'datering', 'stijl',
            'primaire kenmerkgroepen'
        ]
        writer = csv.DictWriter(
            csvfile,
            fieldnames=fieldnames
        )

        writer.writeheader()
        for e in export:
            writer.writerow(e)

def main():

    if CONCEPT:
        logging.info(f'Ophalen van Concept {CONCEPT}')
        erfgoedtypes = get_erfgoedtypes()
        concept = erfgoedtypes.get_by_id(CONCEPT)

        not_concept = []

        for c in NOT_CONCEPT:
            not_concept.append(get_erfgoedtypes().get_by_id(c))

        tpl = 'Concept: %s (%s)'
        logging.info(tpl % (concept.label().label, concept.uri))

        tpl = 'Maar niet concept: %s (%s)'
        for c in not_concept:
            logging.info(tpl % (c.label().label, c.uri))
    else:
        concept = None
        not_concept = []

    logging.info(f'Start genereren CSV')
    generate_csv(
        DISCIPLINE, ERFGOEDWAARDE,
        RECHTSGEVOLGEN,
        concept, not_concept
    )

if __name__ == "__main__":
    main()
